File: src/core/settings_import.py
"""One browser's settings, offered to the server once, with a preview.

THE PREVIEW AND THE IMPORT ARE THE SAME PLAN. ``build_plan`` is called
by both endpoints and ``changes_from`` derives the write from its output,
so the preview cannot describe one thing and the commit perform another.
A preview that lies is worse than no preview: it is a safety control that
tells the user they are safe. That is the one property
``tests/test_settings_import.py`` is built around - it previews, commits,
and asserts the committed values are exactly what the preview named.

THE ALLOWLIST IS A PROJECTION, NOT A SECOND LIST. It is
``ui_preferences.known_fields()`` minus ``REFUSED_FIELDS``, so a field
added to the preference block is importable the day it lands and nobody
has to remember a second place. The two auth tokens never appear here
because they are not preference fields at all, which is a stronger
guarantee than filtering them out: there is no code path on which they
could be named.

ONE FIELD IS REFUSED DESPITE BEING A PREFERENCE, AND THAT IS THE
INTERESTING PART. ``theme_script_consent`` became a shared preference in
#45, which makes it a known field, which would otherwise make it
importable. It must not be. A browser's local record of "I allowed this
theme's script" is the pre-#45 ``cloude.themeJsAllowlist`` shape: it
names a theme id and no digest, so importing it would mint exactly the
unbounded standing grant #45 exists to make unexpressible. The issue's
own rule says it plainly - "never infer theme-script approval from a
theme selection" - and this is the same rule one step further: never
carry a theme-script approval at all. Refusing by NAME here rather than
by hoping nobody adds it to a collector is what makes that true of any
client, including one written later.

SERVER VALUES WIN BY DEFAULT, AND A CONFLICT IS A NAMED OUTCOME. A field
the server already holds is KEPT unless the user selected that specific
field in the preview. There is no "import everything" that quietly wins:
the selection is per field, and a field not in it is reported as
``CONFLICT_KEPT`` rather than dropped silently, so the preview can show
the user both values and which one is about to survive.

THE MARKER IS THE DEFENCE AGAINST A STALE BROWSER. Once an install has
been imported, ``ALREADY_IMPORTED`` is the answer to every later offer.
The failure this prevents is the reason the whole flow is explicit: a
machine that has not been opened in three months connects, uploads its
snapshot, and silently reverts every setting changed since. A marker on
the INSTALL rather than in each browser is what makes one import
one-time across all of them.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, Iterable, List, Optional, Sequence, Tuple

from src.core import ui_preferences

IMPORT_MARKER_KEY = "ui_preferences_import"
"""The top-level config.json key recording that this install has been
imported. A sibling of ``ui_preferences`` rather than a field inside it,
because it describes a one-off migration event and not a preference; a
client hydrating preferences has no business reading it and no reason to
round-trip it."""

MARKER_SCHEMA_VERSION = 1

REFUSED_FIELDS = frozenset({"theme_script_consent"})
"""Preference fields this flow will never carry. See the module
docstring: the local shape of a script consent names no digest, so
importing one would mint an unbounded grant."""

# Per-field outcomes. Every field in a plan carries exactly one.
IMPORTED = "imported"
"""The server held nothing and the candidate will be written."""

IDENTICAL = "identical"
"""The server already holds this exact value. Nothing to write, and
deliberately reported rather than hidden, so the preview's field count
matches what the user can see in their browser."""

CONFLICT_KEPT = "conflict_kept"
"""The server holds a DIFFERENT value and the user did not select this
field. The server's value stands and nothing is written."""

CONFLICT_OVERRIDDEN = "conflict_overridden"
"""The server holds a different value and the user selected this field
in the preview. The candidate will be written."""

REJECTED_INVALID = "rejected_invalid"
"""The candidate failed the preference block's own validation. Nothing
is written for it, and the rest of the plan still stands: one bad value
in a browser must not block the import of the good ones."""

REFUSED = "refused"
"""The field is not importable: not a preference this server knows, or
on ``REFUSED_FIELDS``."""

WRITING_OUTCOMES = frozenset({IMPORTED, CONFLICT_OVERRIDDEN})
"""The only two outcomes that put anything on disk. A caller asks this
set rather than listing the refusals, which can grow."""

# Whole-request outcomes.
COMMITTED = "committed"
UNCHANGED = "unchanged"
ALREADY_IMPORTED = "already_imported"
STALE_REVISION = "stale_revision"


class FieldPlan:
    """What the import will do to one field, and why.

    Attributes:
      field (str): the preference name.
      outcome (str): one of the per-field outcomes above.
      candidate (Any): the value this browser offered, normalised when it
        validated. ``None`` for a refused or unparseable one.
      current (Any): the value the server holds, or ``None`` when it
        holds nothing. ``None`` is genuinely "nothing set" here, because
        the preference block removes a field rather than storing null.
      detail (str|None): a plain sentence for a refusal, safe to show.
    """

    __slots__ = ("field", "outcome", "candidate", "current", "detail")

    def __init__(
        self,
        field: str,
        outcome: str,
        candidate: Any = None,
        current: Any = None,
        detail: Optional[str] = None,
    ) -> None:
        self.field = field
        self.outcome = outcome
        self.candidate = candidate
        self.current = current
        self.detail = detail

    @property
    def writes(self) -> bool:
        """Whether this field ends up on disk.

        Inputs: none.
        Output: bool.
        """
        return self.outcome in WRITING_OUTCOMES

    def as_payload(self) -> Dict[str, Any]:
        """The field plan as plain JSON types, for the preview response.

        Inputs: none.
        Output: dict.
        """
        body: Dict[str, Any] = {
            "field": self.field,
            "outcome": self.outcome,
            "candidate": self.candidate,
            "current": self.current,
            "writes": self.writes,
        }
        if self.detail:
            body["detail"] = self.detail
        return body

    def __repr__(self) -> str:  # pragma: no cover - debug aid only
        return f"<FieldPlan {self.field} {self.outcome}>"


def importable_fields() -> frozenset:
    """Every preference field this flow may carry.

    Description: derived from the preference block rather than listed, so
      a field added there is importable without a second edit here, and
      the refused set is the only thing this module asserts on its own.
    Inputs: none.
    Output: frozenset[str].

    Example:
        'theme' in importable_fields()  -> True
        'theme_script_consent' in importable_fields()  -> False
    """
    return frozenset(ui_preferences.known_fields()) - REFUSED_FIELDS


def build_plan(
    candidates: Dict[str, Any],
    server_values: Dict[str, Any],
    selections: Optional[Sequence[str]] = None,
) -> List[FieldPlan]:
    """Decide, field by field, what an import would do.

    Description: PURE, and the single source of both the preview and the
      write. Fields are planned in a stable sorted order so the preview
      renders the same way twice and a test can compare whole plans.
      Each candidate is validated ON ITS OWN through
      ``ui_preferences.validate_changes``, so one unusable value in a
      browser refuses only itself.
    Inputs:
      candidates (dict) - field name to the value this browser offered.
      server_values (dict) - the preference block's current ``values``.
      selections (sequence[str]|None) - the fields the user explicitly
        chose to overwrite the server with. Anything not named here loses
        to an existing server value.
    Output: list[FieldPlan], one per candidate field, sorted by name.

    Example:
        build_plan({"theme": "matrix"}, {}, None)[0].outcome
        -> 'imported'
    """
    chosen = set(selections or ())
    allowed = importable_fields()
    held = server_values if isinstance(server_values, dict) else {}
    plans: List[FieldPlan] = []

    for field in sorted(candidates.keys()):
        raw = candidates[field]

        if field not in allowed:
            plans.append(
                FieldPlan(
                    field,
                    REFUSED,
                    detail=_refusal_sentence(field),
                )
            )
            continue

        try:
            normalised = ui_preferences.validate_changes({field: raw})[field]
        except ValueError as exc:
            plans.append(
                FieldPlan(field, REJECTED_INVALID, current=held.get(field), detail=str(exc))
            )
            continue

        if normalised is None:
            # The client offered "unset this", which is not something an
            # import has any business doing: it would delete a value the
            # server holds on the strength of a browser having none.
            plans.append(
                FieldPlan(
                    field,
                    REJECTED_INVALID,
                    current=held.get(field),
                    detail="an import can add or replace a setting, never clear one",
                )
            )
            continue

        if field not in held:
            plans.append(FieldPlan(field, IMPORTED, candidate=normalised, current=None))
            continue

        current = held[field]
        if current == normalised:
            plans.append(
                FieldPlan(field, IDENTICAL, candidate=normalised, current=current)
            )
            continue

        outcome = CONFLICT_OVERRIDDEN if field in chosen else CONFLICT_KEPT
        plans.append(
            FieldPlan(field, outcome, candidate=normalised, current=current)
        )

    return plans


def _refusal_sentence(field: str) -> str:
    """Why one field cannot be imported, in a sentence a UI can show.

    Inputs: field (str).
    Output: str.
    """
    if field in REFUSED_FIELDS:
        return (
            "a theme script approval is never imported; this browser's record "
            "names a theme but not the script it approved, so you are asked "
            "again the next time that theme is applied"
        )
    return "this server does not recognise that setting"


def changes_from(plans: Iterable[FieldPlan]) -> Dict[str, Any]:
    """The exact partial update the plan describes.

    Description: THE ONE DERIVATION. The commit writes this and the
      preview renders the plan it came from, so they cannot disagree
      about what is about to happen.
    Inputs: plans (iterable[FieldPlan]).
    Output: dict - field name to value, empty when the plan writes
      nothing.

    Example:
        changes_from(build_plan({"theme": "matrix"}, {}))
        -> {"theme": "matrix"}
    """
    return {plan.field: plan.candidate for plan in plans if plan.writes}


def summarise(plans: Sequence[FieldPlan]) -> Dict[str, int]:
    """Count the plan by outcome, for a one-line summary above the list.

    Inputs: plans (sequence[FieldPlan]).
    Output: dict - outcome name to count, including zeros so a caller
      never has to guess whether a missing key means none or means the
      key was renamed.

    Example:
        summarise(build_plan({"theme": "matrix"}, {}))["imported"] -> 1
    """
    counts = {
        IMPORTED: 0,
        IDENTICAL: 0,
        CONFLICT_KEPT: 0,
        CONFLICT_OVERRIDDEN: 0,
        REJECTED_INVALID: 0,
        REFUSED: 0,
    }
    for plan in plans:
        if plan.outcome in counts:
            counts[plan.outcome] += 1
    return counts


# ---- the completion marker -------------------------------------------


def read_marker(config: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """The import marker stored in a config document, if there is one.

    Description: TOLERANT like ``ui_preferences.read_block``. A marker
      that is not an object reads as absent, because the only thing this
      value gates is whether to OFFER an import, and offering one to a
      user whose marker got mangled is recoverable while refusing one
      forever is not.
    Inputs: config (dict) - the whole parsed config.json.
    Output: dict|None.
    """
    marker = config.get(IMPORT_MARKER_KEY)
    if not isinstance(marker, dict):
        return None
    if not marker.get("completed_at"):
        return None
    return marker


def is_completed(config: Dict[str, Any]) -> bool:
    """Has this install already been imported?

    Inputs: config (dict).
    Output: bool.
    """
    return read_marker(config) is not None


def build_marker(fields: Sequence[str], now: Optional[datetime] = None) -> Dict[str, Any]:
    """The marker to write alongside a successful import.

    Description: records WHICH fields the import actually wrote, not
      which were offered, so a later reader can tell an import that moved
      settings from one that found everything already in place. Both are
      completions and both close the offer.
    Inputs:
      fields (sequence[str]) - the fields written.
      now (datetime|None) - injectable for tests; defaults to UTC now.
    Output: dict.

    Example:
        build_marker(["theme"])["fields"] -> ['theme']
    """
    stamp = (now or datetime.now(timezone.utc)).astimezone(timezone.utc)
    return {
        "schema_version": MARKER_SCHEMA_VERSION,
        "completed_at": stamp.isoformat().replace("+00:00", "Z"),
        "fields": sorted(fields),
    }


def apply_import(
    config: Dict[str, Any],
    changes: Dict[str, Any],
    now: Optional[datetime] = None,
) -> Tuple[Dict[str, Any], Dict[str, Any], Dict[str, Any]]:
    """Merge an import's changes AND its marker into one new document.

    Description: PURE, and ONE document. The values and the marker move
      together or neither moves: a commit that wrote the settings but not
      the marker would leave the install offering the import again, and
      one that wrote the marker but not the settings would close the
      offer having changed nothing. An import that writes NO values still
      writes the marker, because "everything here was already on the
      server" is a completed import and re-offering it helps nobody.
    Inputs:
      config (dict) - the whole parsed config.json, freshly read.
      changes (dict) - the output of ``changes_from``.
      now (datetime|None) - injectable clock.
    Output: (new_config, new_block, moved) - the whole new document, the
      preference block as it now reads, and the fields that actually
      moved.

    Example:
        new_config, block, moved = apply_import(config, {"theme": "matrix"})
    """
    if changes:
        new_config, block, moved = ui_preferences.apply_changes(config, changes)
    else:
        new_config, block, moved = dict(config), ui_preferences.read_block(config), {}
    new_config = dict(new_config)
    new_config[IMPORT_MARKER_KEY] = build_marker(list(moved.keys()), now=now)
    return new_config, block, moved

File: client/js/settings-import.js
/**
 * The one-time "import settings from this browser" control, with its preview.
 *
 * NOTHING IS UPLOADED WITHOUT A PRESS, ON ANY PATH. This module has no
 * load hook, no reconnect hook and no hydration hook. `mount` fetches
 * STATE, which tells it whether the install has already been imported;
 * only then does it read this browser at all, and it reads it to build a
 * PREVIEW, which the server plans and writes nothing for. The commit
 * happens on a click and nowhere else. That is the whole design: an
 * automatic migration means the last browser to connect wins, and a
 * machine nobody has opened in three months would silently revert every
 * setting changed since.
 *
 * THE PREVIEW IS THE SAFETY CONTROL, SO IT SHOWS BOTH VALUES. For a
 * field the server already holds differently, the row shows what this
 * browser has, what the server has, and an unchecked box. Unchecked
 * means the SERVER WINS, which is the default and the direction that
 * cannot lose data the user has not looked at. The commit sends the same
 * candidates and the same selections the preview was built from, and the
 * server rebuilds the plan with the same function, so what is shown is
 * what is done.
 *
 * AN UNAVAILABLE CONTROL EXPLAINS ITSELF RATHER THAN DISAPPEARING. Three
 * states have no import to offer - already imported, nothing recognised
 * in this browser, and everything already matching the server - and each
 * renders a disabled button with the reason beside it. Hiding the
 * control would leave a user looking for something they had been told
 * exists. That is this codebase's convention for an unavailable action,
 * the same one the session row menu follows.
 *
 * THE LOCAL VALUES ARE NEVER CLEARED. Not before the commit, not after
 * it. Browser storage is the fallback every bridged control still reads
 * when the server cannot be reached, and #46's rule is explicit: retain
 * the local source until the server confirms. There is no "and then tidy
 * up" step, because the tidying is what loses a user's settings on the
 * one request that failed.
 */

console.log('[SettingsImport Module] Loading...');

(function () {
    const STATE_ENDPOINT = '/settings/import/state';
    const PREVIEW_ENDPOINT = '/settings/import/preview';
    const COMMIT_ENDPOINT = '/settings/import';

    /** Outcomes the server names for each field. Mirrors settings_import.py. */
    const IMPORTED = 'imported';
    const IDENTICAL = 'identical';
    const CONFLICT_KEPT = 'conflict_kept';
    const CONFLICT_OVERRIDDEN = 'conflict_overridden';
    const REJECTED_INVALID = 'rejected_invalid';
    const REFUSED = 'refused';

    /** What each outcome says on the row, in this app's lowercase voice. */
    const OUTCOME_WORDS = {
        [IMPORTED]: 'will be imported',
        [IDENTICAL]: 'already matches',
        [CONFLICT_KEPT]: 'the server value is kept',
        [CONFLICT_OVERRIDDEN]: 'will replace the server value',
        [REJECTED_INVALID]: 'not a value this server accepts',
        [REFUSED]: 'never imported',
    };

    /** Plain names for the fields, so a row is readable without the schema. */
    const FIELD_WORDS = {
        theme: 'theme',
        audio_enabled: 'sound on or off',
        audio_master_volume: 'sound volume',
        launch_last_model: 'last model chosen',
        sidebar_density: 'sidebar density',
        sidebar_arrangement: 'sidebar pins and order',
        sidebar_pinned: 'sidebar docked',
        config_editor_pinned: 'file drawer docked',
        config_editor_collapsed: 'file tree folds',
        launchpad_collapsed: 'home screen folds',
    };

    let held = null;

    function escapeHtml(str) {
        return String(str === null || str === undefined ? '' : str)
            .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
            .replace(/"/g, '&quot;').replace(/'/g, '&#39;');
    }

    /**
     * Render a stored value in one short readable line.
     *
     * Description: objects are summarised by size rather than dumped, so
     *   a row stays one line and a sidebar arrangement holding 200 names
     *   does not fill the panel. The user is choosing between two
     *   SOURCES here, not auditing two structures.
     * Inputs: value (*).
     * Output: string - already escaped.
     */
    function describe(value) {
        if (value === null || value === undefined) return '<em>not set</em>';
        if (typeof value === 'boolean') return value ? 'on' : 'off';
        if (typeof value === 'number') return escapeHtml(String(value));
        if (typeof value === 'string') {
            return value === '' ? '<em>none</em>' : escapeHtml(value);
        }
        if (Array.isArray(value)) return escapeHtml(value.length + ' entries');
        const keys = Object.keys(value);
        return escapeHtml(keys.length + ' entries');
    }

    function fieldWord(field) {
        return escapeHtml(FIELD_WORDS[field] || field);
    }

    /**
     * Mount the control into a settings slot.
     *
     * Description: fetches state FIRST, so a browser on an
     *   already-imported install never collects anything. Fire and
     *   forget; every failure renders an explanation in place rather
     *   than rejecting into the settings panel's mount pass.
     * Inputs: slot (Element) - the container to render into.
     * Output: undefined.
     * Example: SettingsImport.mount(document.querySelector('#slot'));
     */
    function mount(slot) {
        if (!slot) return;
        slot.innerHTML = shell('<div class="settings-field-hint">reading...</div>');
        refresh(slot);
    }

    async function refresh(slot) {
        const client = globalThis.api;
        if (!client || typeof client.call !== 'function') {
            render(slot, unavailable('the server is not reachable from this page.'));
            return;
        }
        let state;
        try {
            state = await client.call(STATE_ENDPOINT);
        } catch (err) {
            // Deliberately swallowed into an explanation: settings must
            // still open when one panel cannot read its own state.
            console.warn('[SettingsImport] state read failed', err);
            render(slot, unavailable('could not read whether this install has been imported.'));
            return;
        }

        if (state && state.completed) {
            render(slot, completed(state));
            return;
        }

        const collector = globalThis.SettingsImportCollect;
        if (!collector) {
            render(slot, unavailable('this page did not load the settings reader.'));
            return;
        }
        const candidates = collector.collect(globalThis.localStorage, state.importable);
        if (!Object.keys(candidates).length) {
            render(slot, unavailable(
                'this browser has no saved settings to import. that is normal for '
                + 'a browser you have not used the app in before.'
            ));
            return;
        }

        let preview;
        try {
            preview = await client.call(PREVIEW_ENDPOINT, {
                method: 'POST',
                expectedStatuses: [409],
                body: { candidates: candidates, selections: [] },
            });
        } catch (err) {
            console.warn('[SettingsImport] preview failed', err);
            render(slot, unavailable('could not work out what this import would change.'));
            return;
        }

        held = { candidates: candidates, preview: preview, revision: preview.revision };
        render(slot, previewHtml(preview));
        wire(slot);
    }

    function shell(inner) {
        return (
            '<section class="settings-section" data-settings-section="settings-import">'
            + '  <h3 class="settings-section-title">import settings from this browser</h3>'
            + '  <div class="settings-section-description">'
            + '    settings used to be saved in each browser. this moves the ones this '
            + '    browser holds onto the server once, so every device reads the same '
            + '    values. it runs once for this install and nothing is sent until you '
            + '    press import.'
            + '  </div>'
            + inner
            + '</section>'
        );
    }

    function render(slot, inner) {
        slot.innerHTML = shell(inner);
    }

    function unavailable(reason) {
        // Disabled and explained, never hidden. A control that vanishes
        // leaves the user hunting for something they were told exists.
        return (
            '<div class="settings-field-hint">' + escapeHtml(reason) + '</div>'
            + '<button type="button" class="modal-btn modal-btn-secondary" '
            + 'data-settings-import="commit" disabled '
            + 'aria-describedby="settings-import-reason">nothing to import</button>'
        );
    }

    function completed(state) {
        const when = state.completed_at
            ? escapeHtml(String(state.completed_at))
            : 'earlier';
        const fields = (state.imported_fields || []).length;
        return (
            '<div class="settings-field-hint">'
            + 'already imported on this install, ' + when + ', carrying '
            + escapeHtml(String(fields)) + ' setting' + (fields === 1 ? '' : 's') + '. '
            + 'it runs once on purpose: a browser you have not opened in months could '
            + 'otherwise upload its old copy and quietly undo everything changed since. '
            + 'ordinary changes save to the server from now on with no extra step.'
            + '</div>'
            + '<button type="button" class="modal-btn modal-btn-secondary" '
            + 'data-settings-import="commit" disabled>already imported</button>'
        );
    }

    function previewHtml(preview) {
        const rows = (preview.fields || []).map(rowHtml).join('');
        const summary = preview.summary || {};
        const willWrite = (summary[IMPORTED] || 0) + (summary[CONFLICT_OVERRIDDEN] || 0);
        const conflicts = summary[CONFLICT_KEPT] || 0;
        const lines = [
            willWrite + ' setting' + (willWrite === 1 ? '' : 's') + ' will be imported',
        ];
        if (conflicts) {
            lines.push(conflicts + ' already set on the server and kept unless you tick it');
        }
        return (
            '<div class="settings-field-hint" data-settings-import="summary">'
            + escapeHtml(lines.join('. ')) + '.</div>'
            + '<div class="settings-import-rows">' + rows + '</div>'
            + '<button type="button" class="modal-btn modal-btn-primary" '
            + 'data-settings-import="commit"' + (willWrite ? '' : ' disabled')
            + '>' + (willWrite ? 'import these settings' : 'nothing to import') + '</button>'
            + '<div class="settings-field-hint" data-settings-import="status" role="status"></div>'
        );
    }

    function rowHtml(row) {
        const conflict = row.outcome === CONFLICT_KEPT
            || row.outcome === CONFLICT_OVERRIDDEN;
        const box = conflict
            ? ('<input type="checkbox" data-settings-import-field="'
                + escapeHtml(row.field) + '"'
                + (row.outcome === CONFLICT_OVERRIDDEN ? ' checked' : '')
                + ' aria-label="replace the server value for ' + fieldWord(row.field) + '">')
            : '';
        const values = conflict
            ? ('<span class="settings-import-value">this browser: ' + describe(row.candidate)
                + '</span> <span class="settings-import-value">server: '
                + describe(row.current) + '</span>')
            : ('<span class="settings-import-value">' + describe(row.candidate) + '</span>');
        const detail = row.detail
            ? '<div class="settings-field-hint">' + escapeHtml(row.detail) + '</div>'
            : '';
        return (
            '<div class="settings-import-row" data-settings-import-row="'
            + escapeHtml(row.field) + '">'
            + box
            + '<span class="settings-import-name">' + fieldWord(row.field) + '</span>'
            + values
            + '<span class="settings-import-outcome">'
            + escapeHtml(OUTCOME_WORDS[row.outcome] || row.outcome) + '</span>'
            + detail
            + '</div>'
        );
    }

    /**
     * Attach the slot's two delegated listeners, exactly once.
     *
     * Description: `render` replaces the slot's innerHTML but the slot
     *   ELEMENT survives, so a second `addEventListener` on it would not
     *   replace the first - it would stack. After one re-preview every
     *   tick would fire two requests and the import button would commit
     *   twice, which is the one action here that must happen at most
     *   once. The marker lives on the element rather than in module
     *   state because the module is a singleton and the slot is not.
     * Inputs: slot (Element).
     * Output: undefined.
     */
    function wire(slot) {
        if (slot.dataset && slot.dataset.settingsImportWired === '1') return;
        if (slot.dataset) slot.dataset.settingsImportWired = '1';
        slot.addEventListener('change', function (event) {
            const box = event.target.closest
                ? event.target.closest('[data-settings-import-field]') : null;
            if (!box) return;
            reprice(slot);
        });
        slot.addEventListener('click', function (event) {
            const btn = event.target.closest
                ? event.target.closest('[data-settings-import="commit"]') : null;
            if (!btn || btn.disabled) return;
            commit(slot, btn);
        });
    }

    /**
     * Re-plan through the SERVER when a selection changes.
     *
     * Description: the preview is re-fetched rather than adjusted in the
     *   browser. Recomputing it here would mean a second implementation
     *   of the plan, and the whole guarantee of this feature is that
     *   there is exactly one - the moment the displayed plan and the
     *   performed plan come from different code they can differ.
     * Inputs: slot (Element). Output: undefined.
     */
    async function reprice(slot) {
        if (!held) return;
        const selections = currentSelections(slot);
        const client = globalThis.api;
        if (!client || typeof client.call !== 'function') return;
        let preview;
        try {
            preview = await client.call(PREVIEW_ENDPOINT, {
                method: 'POST',
                expectedStatuses: [409],
                body: { candidates: held.candidates, selections: selections },
            });
        } catch (err) {
            console.warn('[SettingsImport] re-preview failed', err);
            return;
        }
        held.preview = preview;
        held.revision = preview.revision;
        render(slot, previewHtml(preview));
        wire(slot);
    }

    function currentSelections(slot) {
        const boxes = slot.querySelectorAll('[data-settings-import-field]');
        const out = [];
        for (let i = 0; i < boxes.length; i += 1) {
            if (boxes[i].checked) out.push(boxes[i].getAttribute('data-settings-import-field'));
        }
        return out;
    }

    async function commit(slot, btn) {
        if (!held) return;
        const statusEl = slot.querySelector('[data-settings-import="status"]');
        const selections = currentSelections(slot);
        btn.disabled = true;
        if (statusEl) statusEl.textContent = 'importing...';

        const client = globalThis.api;
        let body;
        try {
            body = await client.call(COMMIT_ENDPOINT, {
                method: 'POST',
                expectedStatuses: [409],
                body: {
                    candidates: held.candidates,
                    selections: selections,
                    expected_revision: held.revision,
                },
            });
        } catch (err) {
            // A 409 is another device having committed first, or the
            // install having been imported from elsewhere while this
            // panel was open. Both are answers, not faults: nothing was
            // written, this browser's own settings are untouched, and
            // re-reading is what resolves either.
            console.warn('[SettingsImport] import refused', err);
            if (statusEl) {
                statusEl.textContent = 'nothing was imported; this install changed '
                    + 'while the panel was open. your settings in this browser are '
                    + 'unchanged. reopening settings will show what is there now.';
            }
            btn.disabled = false;
            return;
        }

        const moved = Object.keys((body && body.changed) || {}).length;
        if (statusEl) {
            statusEl.textContent = moved
                ? ('imported ' + moved + ' setting' + (moved === 1 ? '' : 's')
                    + '. this browser keeps its own copy of everything it had.')
                : 'everything here already matched the server, so nothing changed.';
        }
        // Re-read the server's own state so what the panel says next is
        // what the server says, not what this browser assumed happened.
        refreshAfterCommit(slot);
    }

    function refreshAfterCommit(slot) {
        const previous = slot.querySelector('[data-settings-import="status"]');
        const message = previous ? previous.textContent : '';
        Promise.resolve(refresh(slot)).then(function () {
            const statusEl = slot.querySelector('[data-settings-import="status"]');
            if (statusEl && message) statusEl.textContent = message;
        }).catch(function (err) {
            console.warn('[SettingsImport] refresh after import failed', err);
        });
    }

    const api = {
        mount: mount,
        // Exported for tests and for a caller that wants the words
        // without the DOM.
        OUTCOME_WORDS: OUTCOME_WORDS,
        FIELD_WORDS: FIELD_WORDS,
    };

    globalThis.SettingsImport = api;
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = api;
    }

    console.log('[SettingsImport Module] Loaded');
})();
